crates/macro-support/src/parser.rs

```diff
@@ -1498,15 +1520,40 @@ impl<'a> MacroParse<(&'a mut TokenStream, BindgenAttrs)> for syn::ItemEnum {
             })
             .collect::<Result<Vec<_>, Diagnostic>>()?;
 
-        let hole = (0..=u32::MAX)
-            .find(|v| !discriminate_map.contains_key(v))
-            .unwrap();
+        if signed {
+            // If the enum has signed discriminants, the entire enum is signed.
+            // This means that some variants with values >= 2^31 will be
+            // invalid, and we have to generate errors for them.
+            // Since it's enough to generate one error, we just check the
+            // variant with the largest value.
+            let max = discriminant_map.iter().max_by_key(|(k, _)| *k).unwrap();
+            if *max.0 > i32::MAX as i64 {
+                bail_span!(
+                    max.1,
+                    "this C-style enums with #[wasm_bindgen] contains at least one negative discriminant, so it can only support numbers that can be represented by `i32`. \
+                    `{}` is too large for `i32`, so either pick a smaller value for this variant or remove all negative discriminants to support values up to `u32::MAX`.",
+                    max.0
+                );
+            }
+        }
+
+        // To make all the code handling holes simpler, we only consider
+        // non-negative holes. This allows us to use `u32` to represent holes.
+        let max = if signed {
+            i32::MAX as i64
+        } else {
+            u32::MAX as i64
+        };
+        let hole = (0..max + 1)
```

Comment:
```suggestion
        let hole = (0..=max)
```

Reply:
It kinda hurts my soul to know that the iteration is less efficient with `..=`, but I agree that this is more readable.

Reply:
Why would it be :open_mouth:?